src/box/tuple_format.c

```diff
@@ -1153,24 +1153,27 @@ tuple_field_map_create_plain(struct tuple_format *format, const char *tuple,
 			 (unsigned) format->exact_field_count);
 		return -1;
 	}
-	defined_field_count = MIN(defined_field_count,
-				  tuple_format_field_count(format));
-
-	void *required_fields = NULL;
-	uint32_t required_fields_sz = BITMAP_SIZE(format->total_field_count);
-	size_t region_svp = region_used(&fiber()->gc);
 
-	if (unlikely(defined_field_count == 0)) {
-		required_fields = format->required_fields;
-		goto end;
+	if (validate && defined_field_count < format->min_field_count) {
+		uint32_t field_count = tuple_format_field_count(format);
+		for (uint32_t i = defined_field_count; i < field_count; i++) {
+			struct tuple_field *field =
+				tuple_format_field(format, i);
+			assert(field != NULL);
+			if (bit_test(format->required_fields, field->id)) {
+				diag_set(ClientError, ER_FIELD_MISSING,
+					 tuple_field_path(field, format));
+				return -1;
+			}
+		}
 	}
 
-	if (validate) {
-		required_fields = xregion_alloc(region, required_fields_sz);
-		memcpy(required_fields, format->required_fields,
-		       required_fields_sz);
-	}
+	defined_field_count = MIN(defined_field_count,
+				  tuple_format_field_count(format));
```

Comment:
Nit: Reusing `defined_field_count` like this here looks confusing: after this statement, the variable doesn't denote the number of fields defined in the tuple anymore.

Reply:
Done.

```diff
--- a/src/box/tuple_format.c
+++ b/src/box/tuple_format.c
@@ -1154,9 +1154,10 @@ tuple_field_map_create_plain(struct tuple_format *format, const char *tuple,
                return -1;
        }

+       uint32_t format_field_count = tuple_format_field_count(format);
        if (validate && defined_field_count < format->min_field_count) {
-               uint32_t field_count = tuple_format_field_count(format);
-               for (uint32_t i = defined_field_count; i < field_count; i++) {
+               for (uint32_t i = defined_field_count;
+                    i < format_field_count; i++) {
                        struct tuple_field *field =
                                tuple_format_field(format, i);
                        assert(field != NULL);
@@ -1168,17 +1169,15 @@ tuple_field_map_create_plain(struct tuple_format *format, const char *tuple,
                }
        }

-       defined_field_count = MIN(defined_field_count,
-                                 tuple_format_field_count(format));
-       if (unlikely(defined_field_count == 0))
+       uint32_t field_count = MIN(defined_field_count, format_field_count);
+       if (unlikely(field_count == 0))
                return 0;

        size_t region_svp = region_used(region);
        struct tuple_field *field;
        struct json_token **token = format->fields.root.children;
        const char *next_pos = pos;
-       for (uint32_t i = 0; i < defined_field_count;
-            i++, token++, pos = next_pos) {
+       for (uint32_t i = 0; i < field_count; i++, token++, pos = next_pos) {
                field = json_tree_entry(*token, struct tuple_field, token);
                bool allow_null = tuple_field_is_nullable(field) ||
                                  tuple_field_has_default(field);
```